File: langgraph-python/app/airline/service.py
from datetime import UTC, datetime
from decimal import Decimal
from uuid import UUID

from app.airline.models import FlightStatus
from app.airline.repository.customer import CustomerRepository
from app.airline.repository.flight import FlightRepository
from app.airline.repository.flight_change import FlightChangeRepository
from app.airline.repository.reservation import ReservationRepository
from app.airline.repository.travel_credit import TravelCreditRepository
from app.airline.schemas import (
    CustomerSchema,
    FlightChangeSchema,
    RebookingOptionSchema,
    ReservationDetailsSchema,
    SearchAvailableFlightsParamsSchema,
    SearchRebookingOptionsInputSchema,
    RebookingResultSchema,
    RebookingSelectionSchema,
    RebookingSnapshotSchema,
    RebookingValidationSchema,
    TravelCreditSchema,
)
from app.airline.repository.errors import RebookingRuleError


class InvalidSearchWindowError(ValueError):
    pass


class InvalidPassengerCountError(ValueError):
    pass


class Service:
    def __init__(
        self,
        customers: CustomerRepository,
        reservations: ReservationRepository,
        flights: FlightRepository,
        credits: TravelCreditRepository,
        changes: FlightChangeRepository,
        rebooking=None,
    ) -> None:
        self._customers = customers
        self._reservations = reservations
        self._flights = flights
        self._credits = credits
        self._changes = changes
        self._rebooking = rebooking

    async def get_customer(self, customer_id: UUID) -> CustomerSchema:
        return CustomerSchema.model_validate(await self._customers.get_by_id(customer_id))

    async def get_reservation(self, booking_reference: str) -> ReservationDetailsSchema:
        if not booking_reference:
            raise ValueError("booking reference is required")
        return await self._reservations.get_details_by_booking_reference(booking_reference)

    async def get_available_travel_credits(
        self, customer_id: UUID, currency: str
    ) -> list[TravelCreditSchema]:
        if not currency:
            raise ValueError("currency is required")
        credits = await self._credits.get_available_by_customer(customer_id, currency)
        return [TravelCreditSchema.model_validate(credit) for credit in credits]

    async def search_rebooking_options(
        self,
        input_data: SearchRebookingOptionsInputSchema,
    ) -> list[RebookingOptionSchema]:
        if input_data.departure_to < input_data.departure_from:
            raise InvalidSearchWindowError("departure window is invalid")
        if input_data.passenger_count <= 0:
            raise InvalidPassengerCountError("passenger count must be greater than zero")

        segment = await self._reservations.get_segment_by_id(input_data.segment_id)
        original_flight = await self._flights.get_by_id(segment.flight_id)
        candidates = await self._flights.search_available(
            SearchAvailableFlightsParamsSchema(
                origin=original_flight.origin_airport,
                destination=original_flight.destination_airport,
                departure_from=input_data.departure_from,
                departure_to=input_data.departure_to,
                passenger_count=input_data.passenger_count,
            )
        )

        original_fare = getattr(segment, "fare_class", None)
        if original_fare is not None and not original_fare.change_allowed:
            return []

        options: list[RebookingOptionSchema] = []
        for candidate in candidates:
            if (
                candidate.flight.status != FlightStatus.SCHEDULED
                or candidate.flight.origin_airport != original_flight.origin_airport
                or candidate.flight.destination_airport != original_flight.destination_airport
                or candidate.flight.departure_at < input_data.departure_from
                or candidate.flight.departure_at > input_data.departure_to
            ):
                continue
            for fare in candidate.fares:
                if (
                    not fare.fare_class.change_allowed
                    or fare.flight_fare.available_seats < input_data.passenger_count
                ):
                    continue
                options.append(
                    RebookingOptionSchema(
                        flight=candidate.flight,
                        fare_class=fare.fare_class,
                        flight_fare=fare.flight_fare,
                        fare_difference=fare.flight_fare.price - segment.price_paid,
                        change_fee=fare.fare_class.change_fee,
                    )
                )
        return options

    async def get_change_history(self, segment_id: UUID) -> list[FlightChangeSchema]:
        changes = await self._changes.get_by_segment(segment_id)
        return [FlightChangeSchema.model_validate(change) for change in changes]

    async def validate_rebooking_selection(
        self, selection: RebookingSelectionSchema
    ) -> RebookingValidationSchema:
        if self._rebooking is None or not hasattr(self._rebooking, "get_snapshot"):
            raise RuntimeError("rebooking repository is required")
        snapshot, credit = await self._rebooking.get_snapshot(selection)
        return validate_rebooking_selection(selection, snapshot, credit)

    async def execute_rebooking(self, selection: RebookingSelectionSchema) -> RebookingResultSchema:
        if self._rebooking is None or not hasattr(self._rebooking, "execute"):
            raise RuntimeError("rebooking repository is required")
        return await self._rebooking.execute(selection, validate_rebooking_selection)

    async def verify_rebooking(self, selection: RebookingSelectionSchema) -> RebookingResultSchema:
        if self._rebooking is None or not hasattr(self._rebooking, "verify"):
            raise RuntimeError("rebooking repository is required")
        return await self._rebooking.verify(selection)


def validate_rebooking_selection(
    selection: RebookingSelectionSchema,
    snapshot: RebookingSnapshotSchema,
    credit: TravelCreditSchema | None,
) -> RebookingValidationSchema:
    if selection.segment_id != snapshot.segment_id:
        raise RebookingRuleError("reservation segment cannot be changed")
    if snapshot.reservation_status.value != "CONFIRMED":
        raise RebookingRuleError("reservation is not confirmed")
    if snapshot.segment_status.value not in {"CONFIRMED", "CHANGED"}:
        raise RebookingRuleError("reservation segment cannot be changed")
    if (
        snapshot.segment_status.value == "CHANGED"
        and snapshot.old_flight_id == selection.new_flight_id
        and snapshot.old_fare_class_id == selection.new_fare_class_id
    ):
        raise RebookingRuleError("rebooking was already applied")
    if not snapshot.old_change_allowed:
        raise RebookingRuleError("fare change is not allowed")
    if (
        snapshot.target_flight_id != selection.new_flight_id
        or snapshot.target_fare_class_id != selection.new_fare_class_id
        or snapshot.target_status.value != "SCHEDULED"
        or snapshot.available_seats < snapshot.passenger_count
    ):
        raise RebookingRuleError("selected flight is unavailable")
    if (
        snapshot.origin_airport != snapshot.target_origin
        or snapshot.destination_airport != snapshot.target_destination
    ):
        raise RebookingRuleError("selected flight route does not match")
    if not snapshot.target_change_allowed:
        raise RebookingRuleError("fare change is not allowed")
    if snapshot.target_currency != snapshot.old_currency:
        raise RebookingRuleError("selected fare currency does not match")
    if snapshot.passenger_count <= 0:
        raise RebookingRuleError("passenger count must be greater than zero")

    amount = selection.travel_credit_amount
    if selection.travel_credit_id is not None:
        if (
            credit is None
            or credit.id != selection.travel_credit_id
            or credit.customer_id != snapshot.customer_id
            or credit.currency != snapshot.old_currency
            or credit.remaining_amount < amount
            or amount <= 0
            or credit.expires_at <= datetime.now(UTC)
            or credit.status.value not in {"AVAILABLE", "PARTIALLY_USED"}
        ):
            raise RebookingRuleError("travel credit is invalid")
    elif amount != 0:
        raise RebookingRuleError("travel credit is invalid")

    difference = snapshot.target_price - snapshot.old_price
    amount_due = max(difference + snapshot.target_change_fee, Decimal("0"))
    if amount_due != amount:
        raise RebookingRuleError("travel credit amount does not match amount due")
    return RebookingValidationSchema(
        selection=selection,
        reservation_id=snapshot.reservation_id,
        customer_id=snapshot.customer_id,
        old_flight_id=snapshot.old_flight_id,
        old_fare_class_id=snapshot.old_fare_class_id,
        new_flight_id=snapshot.target_flight_id,
        new_fare_class_id=snapshot.target_fare_class_id,
        new_price=snapshot.target_price,
        fare_difference=difference,
        change_fee=snapshot.target_change_fee,
        amount_due=amount_due,
        travel_credit_used=amount,
        currency=snapshot.old_currency,
        passenger_count=snapshot.passenger_count,
    )

File: adk-go/internal/airline/service.go
package airline

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

var (
	ErrInvalidSearchWindow     = errors.New("departure window is invalid")
	ErrInvalidPassengerCount   = errors.New("passenger count must be greater than zero")
	ErrFareChangeNotAllowed    = errors.New("fare change is not allowed")
	ErrReservationNotConfirmed = errors.New("reservation is not confirmed")
	ErrSegmentNotChangeable    = errors.New("reservation segment cannot be changed")
	ErrTargetFlightUnavailable = errors.New("selected flight is unavailable")
	ErrRouteMismatch           = errors.New("selected flight route does not match")
	ErrCurrencyMismatch        = errors.New("selected fare currency does not match")
	ErrTravelCreditInvalid     = errors.New("travel credit is invalid")
	ErrRebookingVerification   = errors.New("rebooking verification failed")
	ErrRebookingAlreadyApplied = errors.New("rebooking was already applied")
)

type Service struct {
	logger       *slog.Logger
	customers    CustomerRepository
	reservations ReservationRepository
	flights      FlightRepository
	credits      TravelCreditRepository
	changes      FlightChangeRepository
	rebooking    RebookingRepository
}

type CustomerRepository interface {
	GetByID(context.Context, uuid.UUID) (*Customer, error)
}

type ReservationRepository interface {
	GetDetailsByBookingReference(context.Context, string) (*ReservationDetails, error)
	GetSegmentByID(context.Context, uuid.UUID) (*ReservationSegment, error)
}

type FlightRepository interface {
	GetByID(context.Context, uuid.UUID) (*Flight, error)
	SearchAvailable(context.Context, SearchAvailableFlightsParams) ([]AvailableFlight, error)
}

type TravelCreditRepository interface {
	GetAvailableByCustomer(context.Context, uuid.UUID, string) ([]TravelCredit, error)
}

type FlightChangeRepository interface {
	GetBySegment(context.Context, uuid.UUID) ([]FlightChange, error)
}

type RebookingRepository interface {
	GetSnapshot(context.Context, RebookingSelection) (*RebookingSnapshot, *TravelCredit, error)
	Execute(context.Context, RebookingSelection) (*RebookingResult, error)
	Verify(context.Context, RebookingSelection) (*RebookingResult, error)
}

type AirlineService interface {
	GetAvailableTravelCredits(context.Context, uuid.UUID, string) ([]TravelCredit, error)
	ValidateRebookingSelection(context.Context, RebookingSelection) (RebookingValidation, error)
	ExecuteRebooking(context.Context, RebookingSelection) (*RebookingResult, error)
	VerifyRebooking(context.Context, RebookingSelection) (*RebookingResult, error)
}

func NewService(
	logger *slog.Logger,
	customers CustomerRepository,
	reservations ReservationRepository,
	flights FlightRepository,
	credits TravelCreditRepository,
	changes FlightChangeRepository,
	rebooking ...RebookingRepository,
) *Service {
	var rebookingRepository RebookingRepository
	if len(rebooking) > 0 {
		rebookingRepository = rebooking[0]
	}
	return &Service{
		logger:       logger,
		customers:    customers,
		reservations: reservations,
		flights:      flights,
		credits:      credits,
		changes:      changes,
		rebooking:    rebookingRepository,
	}
}

func ValidateRebookingSelection(selection RebookingSelection, snapshot RebookingSnapshot, credit *TravelCredit) (RebookingValidation, error) {
	if selection.SegmentID == uuid.Nil || snapshot.SegmentID != selection.SegmentID {
		return RebookingValidation{}, ErrSegmentNotChangeable
	}
	if snapshot.ReservationStatus != ReservationStatusConfirmed {
		return RebookingValidation{}, ErrReservationNotConfirmed
	}
	if snapshot.SegmentStatus != SegmentStatusConfirmed && snapshot.SegmentStatus != SegmentStatusChanged {
		return RebookingValidation{}, ErrSegmentNotChangeable
	}
	if snapshot.SegmentStatus == SegmentStatusChanged && snapshot.OldFlightID == selection.NewFlightID && snapshot.OldFareClassID == selection.NewFareClassID {
		return RebookingValidation{}, ErrRebookingAlreadyApplied
	}
	if !snapshot.OldChangeAllowed {
		return RebookingValidation{}, ErrFareChangeNotAllowed
	}
	if snapshot.TargetFlightID != selection.NewFlightID || snapshot.TargetFareClassID != selection.NewFareClassID || snapshot.TargetStatus != FlightStatusScheduled || snapshot.AvailableSeats < 1 {
		return RebookingValidation{}, ErrTargetFlightUnavailable
	}
	if snapshot.OriginAirport != snapshot.TargetOrigin || snapshot.DestinationAirport != snapshot.TargetDestination {
		return RebookingValidation{}, ErrRouteMismatch
	}
	if !snapshot.TargetChangeAllowed {
		return RebookingValidation{}, ErrFareChangeNotAllowed
	}
	if snapshot.TargetCurrency != snapshot.OldCurrency {
		return RebookingValidation{}, ErrCurrencyMismatch
	}
	if snapshot.PassengerCount <= 0 || snapshot.AvailableSeats < snapshot.PassengerCount {
		return RebookingValidation{}, ErrTargetFlightUnavailable
	}
	if selection.TravelCreditID != uuid.Nil {
		if credit == nil || credit.ID != selection.TravelCreditID || credit.CustomerID != snapshot.CustomerID || credit.Currency != snapshot.OldCurrency || credit.RemainingAmount.LessThan(selection.TravelCreditAmount) || selection.TravelCreditAmount.IsNegative() || selection.TravelCreditAmount.IsZero() || !credit.ExpiresAt.After(time.Now()) || (credit.Status != TravelCreditStatusAvailable && credit.Status != TravelCreditStatusPartiallyUsed) {
			return RebookingValidation{}, ErrTravelCreditInvalid
		}
	} else if !selection.TravelCreditAmount.IsZero() {
		return RebookingValidation{}, ErrTravelCreditInvalid
	}
	difference := snapshot.TargetPrice.Sub(snapshot.OldPrice)
	amountDue := difference.Add(snapshot.TargetChangeFee)
	if amountDue.IsNegative() {
		amountDue = decimal.Zero
	}
	if amountDue.GreaterThan(selection.TravelCreditAmount) || selection.TravelCreditAmount.GreaterThan(amountDue) {
		return RebookingValidation{}, ErrTravelCreditInvalid
	}
	return RebookingValidation{Selection: selection, ReservationID: snapshot.ReservationID, CustomerID: snapshot.CustomerID, OldFlightID: snapshot.OldFlightID, OldFareClassID: snapshot.OldFareClassID, NewFlightID: snapshot.TargetFlightID, NewFareClassID: snapshot.TargetFareClassID, NewPrice: snapshot.TargetPrice, FareDifference: difference, ChangeFee: snapshot.TargetChangeFee, AmountDue: amountDue, TravelCreditUsed: selection.TravelCreditAmount, Currency: snapshot.OldCurrency, PassengerCount: snapshot.PassengerCount}, nil
}

func (s *Service) ValidateRebookingSelection(ctx context.Context, selection RebookingSelection) (RebookingValidation, error) {
	s.logger.Debug("validating rebooking selection", "segment_id", selection.SegmentID, "new_flight_id", selection.NewFlightID)
	if s.rebooking == nil {
		return RebookingValidation{}, errors.New("rebooking repository is required")
	}
	snapshot, credit, err := s.rebooking.GetSnapshot(ctx, selection)
	if err != nil {
		return RebookingValidation{}, fmt.Errorf("get rebooking state: %w", err)
	}
	if snapshot == nil {
		return RebookingValidation{}, errors.New("get rebooking state: repository returned nil snapshot")
	}
	return ValidateRebookingSelection(selection, *snapshot, credit)
}

func (s *Service) ExecuteRebooking(ctx context.Context, selection RebookingSelection) (*RebookingResult, error) {
	s.logger.Info("executing rebooking", "segment_id", selection.SegmentID, "new_flight_id", selection.NewFlightID)
	if s.rebooking == nil {
		return nil, errors.New("rebooking repository is required")
	}
	return s.rebooking.Execute(ctx, selection)
}

func (s *Service) VerifyRebooking(ctx context.Context, selection RebookingSelection) (*RebookingResult, error) {
	s.logger.Debug("verifying rebooking", "segment_id", selection.SegmentID, "new_flight_id", selection.NewFlightID)
	if s.rebooking == nil {
		return nil, errors.New("rebooking repository is required")
	}
	return s.rebooking.Verify(ctx, selection)
}

func (s *Service) GetCustomer(ctx context.Context, customerID uuid.UUID) (*Customer, error) {
	return s.customers.GetByID(ctx, customerID)
}

func (s *Service) GetReservation(ctx context.Context, bookingReference string) (*ReservationDetails, error) {
	if bookingReference == "" {
		return nil, errors.New("booking reference is required")
	}
	return s.reservations.GetDetailsByBookingReference(ctx, bookingReference)
}

func (s *Service) GetAvailableTravelCredits(ctx context.Context, customerID uuid.UUID, currency string) ([]TravelCredit, error) {
	if currency == "" {
		return nil, errors.New("currency is required")
	}
	return s.credits.GetAvailableByCustomer(ctx, customerID, currency)
}

func (s *Service) SearchRebookingOptions(ctx context.Context, input SearchRebookingOptionsInput) ([]RebookingOption, error) {
	if input.DepartureTo.Before(input.DepartureFrom) {
		return nil, ErrInvalidSearchWindow
	}
	if input.PassengerCount <= 0 {
		return nil, ErrInvalidPassengerCount
	}
	segment, err := s.reservations.GetSegmentByID(ctx, input.SegmentID)
	if err != nil {
		return nil, fmt.Errorf("get segment for rebooking: %w", err)
	}
	originalFlight, err := s.flights.GetByID(ctx, segment.FlightID)
	if err != nil {
		return nil, fmt.Errorf("get original flight for rebooking: %w", err)
	}
	candidates, err := s.flights.SearchAvailable(ctx, SearchAvailableFlightsParams{
		Origin:         originalFlight.OriginAirport,
		Destination:    originalFlight.DestinationAirport,
		DepartureFrom:  input.DepartureFrom,
		DepartureTo:    input.DepartureTo,
		PassengerCount: input.PassengerCount,
	})
	if err != nil {
		return nil, fmt.Errorf("search rebooking flights: %w", err)
	}

	return buildRebookingOptions(*originalFlight, *segment, candidates, input), nil
}

func buildRebookingOptions(
	originalFlight Flight,
	segment ReservationSegment,
	candidates []AvailableFlight,
	input SearchRebookingOptionsInput,
) []RebookingOption {
	options := make([]RebookingOption, 0)
	originalFareFound := false
	for _, candidate := range candidates {
		for _, fare := range candidate.Fares {
			if fare.FareClass.ID == segment.FareClassID {
				originalFareFound = true
				if !fare.FareClass.ChangeAllowed {
					return options
				}
			}
		}
	}
	if !originalFareFound {
		return options
	}
	for _, candidate := range candidates {
		if candidate.Flight.Status != FlightStatusScheduled ||
			candidate.Flight.OriginAirport != originalFlight.OriginAirport ||
			candidate.Flight.DestinationAirport != originalFlight.DestinationAirport ||
			candidate.Flight.DepartureAt.Before(input.DepartureFrom) ||
			candidate.Flight.DepartureAt.After(input.DepartureTo) {
			continue
		}
		for _, fare := range candidate.Fares {
			if !fare.FareClass.ChangeAllowed || fare.FlightFare.AvailableSeats < input.PassengerCount {
				continue
			}
			options = append(options, RebookingOption{
				Flight:         candidate.Flight,
				FareClass:      fare.FareClass,
				FlightFare:     fare.FlightFare,
				FareDifference: fare.FlightFare.Price.Sub(segment.PricePaid),
				ChangeFee:      fare.FareClass.ChangeFee,
			})
		}
	}
	return options
}

func (s *Service) GetChangeHistory(ctx context.Context, segmentID uuid.UUID) ([]FlightChange, error) {
	return s.changes.GetBySegment(ctx, segmentID)
}
